### ISSUE TITLE
Inconsistent performance between from provided pretrained model weights and published results

### ISSUE THREAD

**Issue by irleader** [Noble-Lab/casanovo#111]
I have used two provided pre-trained model weights provided:
"casanovo_excl_yeast.ckpt" (released on 2022/08/23) and "pretrained_excl_yeast.ckpt" (released on 2022/07/02),
And Casanovo version 3.1.0.

The results I get with eval mode are the same:
      CELoss:valid          0.6026761531829834
   **aa_precision:valid        0.650734770870844**
     aa_recall:valid        0.6479819644090135
    pep_recall:valid        0.4252097971170857

However, from your published paper, the results for using yeast as a test dataset is:
peptide precision: 0.824
peptide precision at coverage 1: 0.561
aa precursion: 0.779
**aa precursion at coverage1: 0.591**

I assume the results from provided pretrained model weights are before applying precursor m/z filter, if yes, why aa_precision is much higher than that from published paper. Casanovo v3.1.0 is still using greedy search, did you make any other improvements on Casanovo after paper publication? Thanks a lot!

**Comment by melihyilmaz** [comment]
Hi,

There hasn't been any changes to the code base that'd improve the performance of the pre-trained models you're using. Which version of yeast data are you using?

p.s. `pep_recall`, which is equivalent to precision at coverage=1, is also different from the results on the ICML paper

**Comment by irleader** [comment]
Hi,

I assume you provided only one version of the 9-species data, there used to be a link on this github page, it is missing now.

The one I am using is "preproc.high.yeast.PXD003868.mgf" with size 529.3MB.

Can you again kindly provide the link to the 9-species data you used for your published data? Thanks!

ps. I also tried other 8 pre-trained models you provided, the results are all different, it might be the problem of the 9-species data I am using.

**Comment by melihyilmaz** [comment]
Here is the [link](https://zenodo.org/record/6791263) to 9-species data used the generate ICML results that was previously on the github page but I think the discrepancy arises because the spectra data on the link is already preprocessed but version 3.1.0 applies preprocessing to spectra by default. You can use an older version of Casanovo like v2.1.1 which has the `preprocess_spec` flag to turn preprocessing off.

Alternatively, you can check out our recent [preprint](https://www.biorxiv.org/content/10.1101/2023.01.03.522621v2) and the accompanying revised [9-species data on MassIVE](https://massive.ucsd.edu/ProteoSAFe/dataset.jsp?accession=MSV000090982) with the latest version of Casanovo.

**Comment by bittremieux** [comment]
There were several functional changes to the code between version 2.1.1, which was used for the ICML paper, and version 3.x.x, also indicated by the major version increase. Hence also the boost in performance. 🙂

If you want to reproduce the ICML results, you'll have to use Casanovo version 2.1.1. For the results described in our recent preprint, use the latest version and the new weights released with Casanovo v3.0.0.

**Comment by irleader** [comment]
Thanks for your information.

Following your advice, to reproduce the results in ICML, I used Casanovo v2.1.1 (with preprocess_spec: False), "casanovo_excl_yeast.ckpt" (released on 2022/08/23) and "pretrained_excl_yeast.ckpt" (released on 2022/07/02), with the preprocessed "preproc.high.yeast.PXD003868.mgf" in the link provided.

The results seem to be more interesting, both pretrained weights still give the same performance, but better than that from Casanovo v3.1.0:

      CELoss:valid          0.5459600687026978
   **aa_precision:valid       0.7046437859535217**
     aa_recall:valid        0.7060638070106506
    pep_recall:valid        0.5047619938850403

I am not sure what might went wrong. If this is not important, I will instead try to reproduce the results in the preprint. While there are only graphs in the preprint, is it possible to provide the exact precision/recall numbers from the nine-species experiments with casanovo v2.1.1 and v3.2.0, the same table from ICML paper. Such that I can refer. Thanks a lot.

**Comment by bittremieux** [comment]
Weights files "casanovo_excl_yeast.ckpt" and "pretrained_excl_yeast.ckpt" are identical, so they should indeed give the same results. I do get the same precision and recall as you reported when using Casanovo v2.1.1 with those weights, but I'm not entirely sure why the results deviate from what is reported in the manuscript.

The performance difference between Casanovo v2.1.1 and Casanovo v3.2.0 on this yeast dataset likely stems from the fact that in our initial paper we used the peptide labels from the original benchmark by Tran et al. (DeepNovo paper). However, we noticed that search settings to obtain those PSMs were suboptimal and redid the benchmark. Consequently, Casanovo v2.1.1 learned some of those inaccuracies (e.g. deamidation vs first isotopic peak), whereas Casanovo v3.x.x should give better results, even though some of those are erroneously deemed incorrect in the original version of the benchmark. For more details, see the second Results subsection and the "Nine-species benchmark dataset" section in the Methods in the preprint.

@melihyilmaz Can you share the numeric values for the updated 9-species benchmark (Figure S5)?

**Comment by irleader** [comment]
Hi,

Thanks a lot for your clarification.

By applying the original nine species dataset from DeepNovo(Tran et al.) and your preprocessed one based on the original dataset, together with correct preprocess specs, I am able to obtain consistent performances with casanovo v2.1.1(preprocess_spec False with preprocessed dataset), v2.1.1 (preprocess_spec True with deepnovo original raw dataset) and v3.1.0 (with preprocessed dataset) as follows:

      CELoss:valid          0.5461629033088684
   aa_precision:valid       0.7057459339992785
     aa_recall:valid        0.7071716086435299
    pep_recall:valid        0.5046985050978404

Although they are still not the same from those from ICML manuscript, they show the consistence among different versions of casanovo.

Next step, I will focus on your revised [9-species raw data,](https://massive.ucsd.edu/ProteoSAFe/dataset.jsp?accession=MSV000090982) and use it to benchmark both casanovo v3.1.0 (greedy search)  and v3.2.0 (beam search). I am very looking forward to getting your pretrained weights and numerical value of precision/recall on your revised 9-species dataset. Thanks.

**Comment by bittremieux** [comment]
You can get similar performance to beam search with Casanovo v3.2.0 by setting the [number of beams](https://github.com/Noble-Lab/casanovo/blob/main/casanovo/config.yaml#L68) to 1. For Casanovo v3+ the recommended weights are those from training on the larger MassIVE-KB dataset, rather than separate weights for the (smaller) 9-species data.

**Comment by irleader** [comment]
Yes, it gets the same results with casanovo v3.2.0 by setting beam=1.

By looking into the codes, it seems precursor m/z filter is not applied to any versions of casanovo, am I right?

So if I would like to calculate AA/peptide precision/recall after applying precursor m/z filter , I need to use 'denovo' mode to get the predicted output file first, and manually exclude those predictions with precursor mass difference greater than ε (considering isotope error range between 0 and 1).

As described in preprint, "The threshold value ϵ is a property of the mass spectrometer that the data is collected with, and hence is known at inference time. Accordingly, we choose ϵ based on the precursor mass error tolerance used in the database search to obtain ground truth peptide sequences for the test data."  So for original/revised 9-species-data, ϵ is the second last column "precursor" in Table S2 (ranging from 10 to 30 ppm for different species)? Or should it all be 50ppm( which is specified by "precursor_mass_tol" in config.yaml)? Thanks!

**Comment by wsnoble** [comment]
As of version 3.0.0, the precursor m/z filter is implemented in Casanovo (see the [changelog](https://github.com/Noble-Lab/casanovo/blob/main/CHANGELOG.md)).

You can set the precursor window size via the precursor_mass_tol parameter in the [config file](https://github.com/Noble-Lab/casanovo/blob/main/casanovo/config.yaml).
